File: Labs/Lab 6/Solution.java
import java.util.*;

public class Solution{
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        List<List<Integer>> adjList = new ArrayList<>();
        for(int i = 0; i<numCourses; i++){
            adjList.add(new ArrayList<>());
        }

        for(int[] prereq:prerequisites){
            adjList.get(prereq[1]).add(prereq[0]);
        }

        int[] visited = new int[numCourses];

        for(int i = 0; i<numCourses;i++){
            if(visited[i] == 0 && !dfs(adjList,visited, i)) return false; // cycle
        }
        return true;
    }

    private boolean dfs(List<List<Integer>> adjList, int[] visited, int course){
        visited[course] = 1;
        for(int nextCourse:adjList.get(course)){
            if(visited[nextCourse]==1) return false; //cycle
            if(visited[nextCourse]==0 && !dfs(adjList, visited, nextCourse)) return false; // cycle in subtree
        }
        visited[course]=2;
        return true;    
    }

    public static void main(String[] args) {
        Solution schedule = new Solution();

        // Test cases
        System.out.println(schedule.canFinish(2, new int[][]{{1, 0}})); //Case 1
        System.out.println(schedule.canFinish(2, new int[][]{{1, 0}, {0, 1}})); //Case 2
    }
}

File: Labs/Lab 6/Solution.py
class Solution:
    @staticmethod
    def canFinish(numCourses, prerequisites):
        adjList = [[] for _ in range (numCourses)]
        for prereq in prerequisites:
            adjList[prereq[1]].append(prereq[0])

        visited = [0] * numCourses
        def dfs(course): #put it inside, so less parameters
            visited[course] = 1
            for nextCourse in adjList[course]:
                if visited[nextCourse] == 1: return False
                if visited[nextCourse] == 0 and not dfs(nextCourse):
                    return False # cycle in subtree
            visited[course]=2
            return True

        for i in range(numCourses):
            if visited[i] == 0 and not dfs(i): 
                return False #cycle
            
        return True
    
schedule = Solution()
print(schedule.canFinish(2, [[1,0]]))
print(schedule.canFinish(2,[[1,0], [0,1]]))
        